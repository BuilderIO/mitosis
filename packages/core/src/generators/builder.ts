import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { BuilderElement } from '@builder.io/sdk';
import { getStateObjectString } from '../helpers/get-state-object-string';
import { fastClone } from '../helpers/fast-clone';
import dedent from 'dedent';
import { format } from 'prettier/standalone';
import json5 from 'json5';
import { isUpperCase } from '../helpers/is-upper-case';
import { mediaQueryRegex, sizes } from '../constants/media-sizes';
import { filterEmptyTextNodes } from '../helpers/filter-empty-text-nodes';
import { isComponent } from '../helpers/is-component';

const builderBlockPrefixes = ['Amp', 'Core', 'Builder', 'Raw', 'Form'];
const mapComponentName = (name: string) => {
  if (name === 'CustomCode') {
    return 'Custom Code';
  }
  for (const prefix of builderBlockPrefixes) {
    if (name.startsWith(prefix)) {
      const suffix = name.replace(prefix, '');
      if (isUpperCase(suffix[0])) {
        return `${prefix}:${name.replace(prefix, '')}`;
      }
    }
  }
  return name;
};

const componentMappers: {
  [key: string]: (
    node: JSXLiteNode,
    options: ToBuilderOptions,
  ) => BuilderElement;
} = {
  Columns(node, options) {
    const block = blockToBuilder(node, options, { skipMapper: true });

    const columns = block.children!.map((item) => ({
      blocks: item.children,
    }));

    block.component!.options.columns = columns;

    block.children = [];

    return block;
  },
  For(node, options) {
    return el(
      {
        component: {
          name: 'Fragment',
        },
        repeat: {
          collection: node.bindings.each as string,
          itemName: node.bindings._forName as string,
        },
        children: node.children
          .filter(filterEmptyTextNodes)
          .map((node) => blockToBuilder(node, options)),
      },
      options,
    );
  },
  Show(node, options) {
    return el(
      {
        // TODO: the reverse mapping for this
        component: {
          name: 'Fragment',
        },
        bindings: {
          show: node.bindings.when as string,
        },
        children: node.children
          .filter(filterEmptyTextNodes)
          .map((node) => blockToBuilder(node, options)),
      },
      options,
    );
  },
};

const el = (
  options: Partial<BuilderElement>,
  toBuilderOptions: ToBuilderOptions,
): BuilderElement => ({
  '@type': '@builder.io/sdk:Element',
  ...(toBuilderOptions.includeIds && {
    id:
      'builder-' +
      Math.random()
        .toString(36)
        .split('.')[1],
  }),
  ...options,
});

export type ToBuilderOptions = {
  includeIds?: boolean;
};

function tryFormat(code: string) {
  let str = code;
  try {
    str = format(str, {
      parser: 'babel',
      plugins: [
        require('prettier/parser-babel'), // To support running in browsers
      ],
    });
  } catch (err) {
    console.error('Format error for code:', str);
    throw err;
  }
  return str;
}

type InternalOptions = {
  skipMapper?: boolean;
};

export const blockToBuilder = (
  json: JSXLiteNode,
  options: ToBuilderOptions = {},
  _internalOptions: InternalOptions = {},
): BuilderElement => {
  const mapper = !_internalOptions.skipMapper && componentMappers[json.name];
  if (mapper) {
    return mapper(json, options);
  }
  if (json.properties._text || json.bindings._text) {
    return el(
      {
        tagName: 'span',
        // responsiveStyles: {
        //   large: json.properties.css as any,
        // },
        bindings: {
          // TODO: css to responsiveStyles and back
          // ...(json.bindings as any),
          ...(json.bindings._text
            ? {
                'component.options.text': json.bindings._text as string,
                'json.bindings._text': undefined as any,
              }
            : {}),
        },
        component: {
          name: 'Text',
          options: {
            text: json.properties._text,
          },
        },
      },
      options,
    );
  }

  const thisIsComponent = isComponent(json);

  let bindings = thisIsComponent ? {} : json.bindings;
  if (thisIsComponent) {
    for (const key in json.bindings) {
      bindings[`component.options.${key}`] = json.bindings[key];
    }
  }

  const hasCss = !!json.bindings.css;
  let responsiveStyles: {
    large: Partial<CSSStyleDeclaration>;
    medium?: Partial<CSSStyleDeclaration>;
    small?: Partial<CSSStyleDeclaration>;
  } = {
    large: {},
  };
  if (hasCss) {
    const cssRules = json5.parse(json.bindings.css as string);
    const cssRuleKeys = Object.keys(cssRules);
    for (const ruleKey of cssRuleKeys) {
      const mediaQueryMatch = ruleKey.match(mediaQueryRegex);
      if (mediaQueryMatch) {
        const [fullmatch, pixelSize] = mediaQueryMatch;
        const sizeForWidth = sizes.getSizeForWidth(Number(pixelSize));
        const currentSizeStyles = responsiveStyles[sizeForWidth] || {};
        responsiveStyles[sizeForWidth] = {
          ...currentSizeStyles,
          ...cssRules[ruleKey],
        };
      } else {
        responsiveStyles.large = {
          ...responsiveStyles.large,
          [ruleKey]: cssRules[ruleKey],
        };
      }
    }
  }

  return el(
    {
      tagName: thisIsComponent ? undefined : json.name,
      ...(hasCss && {
        responsiveStyles,
      }),
      ...(thisIsComponent && {
        component: {
          name: mapComponentName(json.name),
          options: json.properties,
        },
      }),
      properties: thisIsComponent ? undefined : (json.properties as any),
      bindings: thisIsComponent ? undefined : (json.bindings as any),
      children: json.children
        .filter(filterEmptyTextNodes)
        .map((child) => blockToBuilder(child, options)),
    },
    options,
  );
};

export const componentToBuilder = (
  componentJson: JSXLiteComponent,
  options: ToBuilderOptions = {},
) => {
  return fastClone({
    data: {
      jsCode: tryFormat(dedent`
        Object.assign(state, ${getStateObjectString(componentJson)});
      `),
      blocks: componentJson.children
        .filter(filterEmptyTextNodes)
        .map((child) => blockToBuilder(child, options)),
    },
  });
};
