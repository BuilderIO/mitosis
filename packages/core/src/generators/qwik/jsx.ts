import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
