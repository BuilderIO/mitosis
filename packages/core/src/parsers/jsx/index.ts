export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
