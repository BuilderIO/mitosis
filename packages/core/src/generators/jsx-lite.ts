import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { fastClone } from '../helpers/fast-clone';
import { getComponents } from '../helpers/get-components';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectString } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { selfClosingTags } from '../parsers/jsx';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

type ToJsxLiteOptions = {
  prettier?: boolean;
};
export const blockToJsxLite = (
  json: JSXLiteNode,
  options: ToJsxLiteOptions = {},
): string => {
  if (json.name === 'For') {
    const needsWrapper = json.children.length !== 1;
    return `<For each={${json.bindings.each}}>
    {(${json.bindings._forName}, index) =>
      ${needsWrapper ? '<>' : ''}
        ${json.children.map((child) => blockToJsxLite(child, options))}}
      ${needsWrapper ? '</>' : ''}
    </For>`;
  }

  if (json.properties._text) {
    return json.properties._text as string;
  }

  if (json.bindings._text) {
    return `{${json.bindings._text}}`;
  }

  let str = '';

  str += `<${json.name} `;

  if (json.bindings._spread) {
    str += ` {...(${json.bindings._spread})} `;
  }

  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    if (key === '_spread') {
      continue;
    }

    if (key.startsWith('on')) {
      str += ` ${key}={event => (${value})} `;
    } else {
      str += ` ${key}={${value}} `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children
      .map((item) => blockToJsxLite(item, options))
      .join('\n');
  }

  str += `</${json.name}>`;

  return str;
};

const getRefsString = (json: JSXLiteComponent, refs = getRefs(json)) => {
  let str = '';

  for (const ref of Array.from(refs)) {
    str += `\nconst ${ref} = useRef();`;
  }

  return str;
};

const jsxLiteCoreComponents = ['Show', 'For'];

export const componentToJsxLite = (
  componentJson: JSXLiteComponent,
  options: ToJsxLiteOptions = {},
) => {
  const json = fastClone(componentJson);

  const refs = getRefs(json);

  mapRefs(json, (refName) => `${refName}.current`);

  const addWrapper = json.children.length > 1;

  const components = Array.from(getComponents(json));

  const jsxLiteComponents = components.filter((item) =>
    jsxLiteCoreComponents.includes(item),
  );
  const otherComponents = components.filter(
    (item) => !jsxLiteCoreComponents.includes(item),
  );

  const hasState = Boolean(Object.keys(componentJson.state).length);

  const needsJsxLiteCoreImport = Boolean(
    hasState || refs.size || jsxLiteComponents.length,
  );

  // TODO: smart only pull in imports as needed
  let str = dedent`
    ${
      !needsJsxLiteCoreImport
        ? ''
        : `import { ${!hasState ? '' : 'useState, '} ${
            !refs.size ? '' : 'useRef, '
          } ${jsxLiteComponents.join(', ')} } from '@jsx-lite/core';`
    }
    ${
      !otherComponents.length
        ? ''
        : `import { ${otherComponents.join(
            ',',
          )} } from '@builder.io/components';`
    }
    ${renderPreComponent(json)}

    export default function MyComponent(props) {
      ${
        !hasState
          ? ''
          : `const state = useState(${getStateObjectString(json)});`
      }
      ${getRefsString(json)}

      return (${addWrapper ? '<div>' : ''}
        ${json.children.map((item) => blockToJsxLite(item)).join('\n')}
        ${addWrapper ? '</div>' : ''})
    }

  `;

  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'typescript',
        plugins: [
          require('prettier/parser-typescript'), // To support running in browsers
        ],
      });
    } catch (err) {
      console.error(
        'Format error for file:',
        str,
        JSON.stringify(json, null, 2),
      );
      throw err;
    }
  }
  return str;
};
