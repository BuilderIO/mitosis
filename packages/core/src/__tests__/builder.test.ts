import { componentToBuilder } from '../generators/builder';
import { componentToJsxLite } from '../generators/jsx-lite';
import { builderContentToJsxLiteComponent } from '../parsers/builder';
import { parseJsx } from '../parsers/jsx';

const stamped = require('./data/blocks/stamped-io.raw');
const customCode = require('./data/blocks/custom-code.raw');
const embed = require('./data/blocks/embed.raw');

describe('Builder', () => {
  test('Stamped', () => {
    const json = parseJsx(stamped);
    const builderJson = componentToBuilder(json);
    expect(builderJson).toMatchSnapshot();

    const backToJsxLite = builderContentToJsxLiteComponent(builderJson);
    const jsxLite = componentToJsxLite(backToJsxLite);
    expect(jsxLite).toMatchSnapshot();
  });

  test('CustomCode', () => {
    const json = parseJsx(customCode);
    const builderJson = componentToBuilder(json);
    expect(builderJson).toMatchSnapshot();

    const backToJsxLite = builderContentToJsxLiteComponent(builderJson);
    const jsxLite = componentToJsxLite(backToJsxLite);
    expect(jsxLite).toMatchSnapshot();
  });

  test('Embed', () => {
    const json = parseJsx(embed);
    const builderJson = componentToBuilder(json);
    expect(builderJson).toMatchSnapshot();

    const backToJsxLite = builderContentToJsxLiteComponent(builderJson);
    const jsxLite = componentToJsxLite(backToJsxLite);
    expect(jsxLite).toMatchSnapshot();
  });
});
