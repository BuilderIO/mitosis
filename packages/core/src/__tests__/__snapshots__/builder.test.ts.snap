// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder CustomCode 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "@type": "@builder.io/sdk:Element",
        "actions": Object {},
        "bindings": Object {
          "className": "'builder-custom-code' + (this.props.replaceNodes ? ' replace-nodes' : '')",
          "innerHTML": "props.code",
          "ref": "elem",
        },
        "children": Array [],
        "code": Object {
          "actions": Object {},
        },
        "properties": Object {},
        "tagName": "div",
      },
    ],
    "jsCode": "var props = state;

Object.assign(state, {
  scriptsInserted: [],
  scriptsRun: [],
  findAndRunScripts() {
    // TODO: Move this function to standalone one in '@builder.io/utils'
    if (elem && typeof window !== \\"undefined\\") {
      const scripts = elem.getElementsByTagName(\\"script\\");

      for (let i = 0; i < scripts.length; i++) {
        const script = scripts[i];

        if (script.src) {
          if (state.scriptsInserted.includes(script.src)) {
            continue;
          }

          state.scriptsInserted.push(script.src);
          const newScript = document.createElement(\\"script\\");
          newScript.async = true;
          newScript.src = script.src;
          document.head.appendChild(newScript);
        } else if (
          !script.type ||
          [
            \\"text/javascript\\",
            \\"application/javascript\\",
            \\"application/ecmascript\\",
          ].includes(script.type)
        ) {
          if (state.scriptsRun.includes(script.innerText)) {
            continue;
          }

          try {
            state.scriptsRun.push(script.innerText);
            new Function(script.innerText)();
          } catch (error) {
            console.warn(\\"\`CustomCode\`: Error running script:\\", error);
          }
        }
      }
    }
  },
});

state.findAndRunScripts();
",
    "tsCode": "var props = state;

useState({
  scriptsInserted: [],
  scriptsRun: [],
  findAndRunScripts() {
    // TODO: Move this function to standalone one in '@builder.io/utils'
    if (elem && typeof window !== \\"undefined\\") {
      const scripts = elem.getElementsByTagName(\\"script\\");

      for (let i = 0; i < scripts.length; i++) {
        const script = scripts[i];

        if (script.src) {
          if (state.scriptsInserted.includes(script.src)) {
            continue;
          }

          state.scriptsInserted.push(script.src);
          const newScript = document.createElement(\\"script\\");
          newScript.async = true;
          newScript.src = script.src;
          document.head.appendChild(newScript);
        } else if (
          !script.type ||
          [
            \\"text/javascript\\",
            \\"application/javascript\\",
            \\"application/ecmascript\\",
          ].includes(script.type)
        ) {
          if (state.scriptsRun.includes(script.innerText)) {
            continue;
          }

          try {
            state.scriptsRun.push(script.innerText);
            new Function(script.innerText)();
          } catch (error) {
            console.warn(\\"\`CustomCode\`: Error running script:\\", error);
          }
        }
      }
    }
  },
});

onMount(() => {
  state.findAndRunScripts();
});
",
  },
}
`;

exports[`Builder CustomCode 2`] = `
"import { useRef } from \\"@jsx-lite/core\\";

export default function MyComponent(props) {
  const elem = useRef();

  onMount(() => {
    useState({
      scriptsInserted: [],
      scriptsRun: [],

      findAndRunScripts() {
        // TODO: Move this function to standalone one in '@builder.io/utils'
        if (elem.current && typeof window !== \\"undefined\\") {
          const scripts = elem.current.getElementsByTagName(\\"script\\");

          for (let i = 0; i < scripts.length; i++) {
            const script = scripts[i];

            if (script.src) {
              if (state.scriptsInserted.includes(script.src)) {
                continue;
              }

              state.scriptsInserted.push(script.src);
              const newScript = document.createElement(\\"script\\");
              newScript.async = true;
              newScript.src = script.src;
              document.head.appendChild(newScript);
            } else if (
              !script.type ||
              [
                \\"text/javascript\\",
                \\"application/javascript\\",
                \\"application/ecmascript\\",
              ].includes(script.type)
            ) {
              if (state.scriptsRun.includes(script.innerText)) {
                continue;
              }

              try {
                state.scriptsRun.push(script.innerText);
                new Function(script.innerText)();
              } catch (error) {
                console.warn(\\"\`CustomCode\`: Error running script:\\", error);
              }
            }
          }
        }
      },
    });
    onMount(() => {
      state.findAndRunScripts();
    });
  });

  return (
    <div
      ref={elem}
      className={
        \\"builder-custom-code\\" +
        (this.props.replaceNodes ? \\" replace-nodes\\" : \\"\\")
      }
      innerHTML={props.code}
    ></div>
  );
}
"
`;

exports[`Builder Embed 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "@type": "@builder.io/sdk:Element",
        "actions": Object {},
        "bindings": Object {
          "innerHTML": "props.content",
          "ref": "elem",
        },
        "children": Array [],
        "code": Object {
          "actions": Object {},
        },
        "properties": Object {
          "className": "builder-embed",
        },
        "tagName": "div",
      },
    ],
    "jsCode": "var props = state;

Object.assign(state, {
  scriptsInserted: [],
  scriptsRun: [],
  findAndRunScripts() {
    // TODO: Move this function to standalone one in '@builder.io/utils'
    if (elem && typeof window !== \\"undefined\\") {
      const scripts = elem.getElementsByTagName(\\"script\\");

      for (let i = 0; i < scripts.length; i++) {
        const script = scripts[i];

        if (script.src) {
          if (state.scriptsInserted.includes(script.src)) {
            continue;
          }

          state.scriptsInserted.push(script.src);
          const newScript = document.createElement(\\"script\\");
          newScript.async = true;
          newScript.src = script.src;
          document.head.appendChild(newScript);
        } else if (
          !script.type ||
          [
            \\"text/javascript\\",
            \\"application/javascript\\",
            \\"application/ecmascript\\",
          ].includes(script.type)
        ) {
          if (state.scriptsRun.includes(script.innerText)) {
            continue;
          }

          try {
            state.scriptsRun.push(script.innerText);
            new Function(script.innerText)();
          } catch (error) {
            console.warn(\\"\`Embed\`: Error running script:\\", error);
          }
        }
      }
    }
  },
});

state.findAndRunScripts();
",
    "tsCode": "var props = state;

useState({
  scriptsInserted: [],
  scriptsRun: [],
  findAndRunScripts() {
    // TODO: Move this function to standalone one in '@builder.io/utils'
    if (elem && typeof window !== \\"undefined\\") {
      const scripts = elem.getElementsByTagName(\\"script\\");

      for (let i = 0; i < scripts.length; i++) {
        const script = scripts[i];

        if (script.src) {
          if (state.scriptsInserted.includes(script.src)) {
            continue;
          }

          state.scriptsInserted.push(script.src);
          const newScript = document.createElement(\\"script\\");
          newScript.async = true;
          newScript.src = script.src;
          document.head.appendChild(newScript);
        } else if (
          !script.type ||
          [
            \\"text/javascript\\",
            \\"application/javascript\\",
            \\"application/ecmascript\\",
          ].includes(script.type)
        ) {
          if (state.scriptsRun.includes(script.innerText)) {
            continue;
          }

          try {
            state.scriptsRun.push(script.innerText);
            new Function(script.innerText)();
          } catch (error) {
            console.warn(\\"\`Embed\`: Error running script:\\", error);
          }
        }
      }
    }
  },
});

onMount(() => {
  state.findAndRunScripts();
});
",
  },
}
`;

exports[`Builder Embed 2`] = `
"import { useRef } from \\"@jsx-lite/core\\";

export default function MyComponent(props) {
  const elem = useRef();

  onMount(() => {
    useState({
      scriptsInserted: [],
      scriptsRun: [],

      findAndRunScripts() {
        // TODO: Move this function to standalone one in '@builder.io/utils'
        if (elem.current && typeof window !== \\"undefined\\") {
          const scripts = elem.current.getElementsByTagName(\\"script\\");

          for (let i = 0; i < scripts.length; i++) {
            const script = scripts[i];

            if (script.src) {
              if (state.scriptsInserted.includes(script.src)) {
                continue;
              }

              state.scriptsInserted.push(script.src);
              const newScript = document.createElement(\\"script\\");
              newScript.async = true;
              newScript.src = script.src;
              document.head.appendChild(newScript);
            } else if (
              !script.type ||
              [
                \\"text/javascript\\",
                \\"application/javascript\\",
                \\"application/ecmascript\\",
              ].includes(script.type)
            ) {
              if (state.scriptsRun.includes(script.innerText)) {
                continue;
              }

              try {
                state.scriptsRun.push(script.innerText);
                new Function(script.innerText)();
              } catch (error) {
                console.warn(\\"\`Embed\`: Error running script:\\", error);
              }
            }
          }
        }
      },
    });
    onMount(() => {
      state.findAndRunScripts();
    });
  });

  return (
    <div className=\\"builder-embed\\" ref={elem} innerHTML={props.content}></div>
  );
}
"
`;

exports[`Builder Stamped 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "@type": "@builder.io/sdk:Element",
        "actions": Object {},
        "bindings": Object {},
        "children": Array [
          Object {
            "@type": "@builder.io/sdk:Element",
            "actions": Object {
              "click": "state.showReviewPrompt = true",
            },
            "bindings": Object {},
            "children": Array [
              Object {
                "@type": "@builder.io/sdk:Element",
                "bindings": Object {},
                "component": Object {
                  "name": "Text",
                  "options": Object {
                    "text": "
        Write a review
      ",
                  },
                },
                "tagName": "span",
              },
            ],
            "code": Object {
              "actions": Object {
                "click": "state.showReviewPrompt = true",
              },
            },
            "properties": Object {},
            "tagName": "button",
          },
          Object {
            "@type": "@builder.io/sdk:Element",
            "bindings": Object {
              "show": "state.showReviewPrompt",
            },
            "children": Array [
              Object {
                "@type": "@builder.io/sdk:Element",
                "actions": Object {},
                "bindings": Object {},
                "children": Array [],
                "code": Object {
                  "actions": Object {},
                },
                "properties": Object {
                  "placeholder": "Email",
                },
                "tagName": "input",
              },
              Object {
                "@type": "@builder.io/sdk:Element",
                "actions": Object {},
                "bindings": Object {},
                "children": Array [],
                "code": Object {
                  "actions": Object {},
                },
                "properties": Object {
                  "placeholder": "Title",
                },
                "responsiveStyles": Object {
                  "large": Object {
                    "display": "block",
                  },
                },
                "tagName": "input",
              },
              Object {
                "@type": "@builder.io/sdk:Element",
                "actions": Object {},
                "bindings": Object {},
                "children": Array [],
                "code": Object {
                  "actions": Object {},
                },
                "properties": Object {
                  "placeholder": "How was your experience?",
                },
                "responsiveStyles": Object {
                  "large": Object {
                    "display": "block",
                  },
                },
                "tagName": "textarea",
              },
              Object {
                "@type": "@builder.io/sdk:Element",
                "actions": Object {
                  "click": "state.showReviewPrompt = false",
                },
                "bindings": Object {},
                "children": Array [
                  Object {
                    "@type": "@builder.io/sdk:Element",
                    "bindings": Object {},
                    "component": Object {
                      "name": "Text",
                      "options": Object {
                        "text": "
          Submit
        ",
                      },
                    },
                    "tagName": "span",
                  },
                ],
                "code": Object {
                  "actions": Object {
                    "click": "state.showReviewPrompt = false",
                  },
                },
                "properties": Object {},
                "responsiveStyles": Object {
                  "large": Object {
                    "display": "block",
                  },
                },
                "tagName": "button",
              },
            ],
            "component": Object {
              "name": "Fragment",
            },
          },
          Object {
            "@type": "@builder.io/sdk:Element",
            "children": Array [
              Object {
                "@type": "@builder.io/sdk:Element",
                "actions": Object {},
                "bindings": Object {},
                "children": Array [
                  Object {
                    "@type": "@builder.io/sdk:Element",
                    "actions": Object {},
                    "bindings": Object {
                      "src": "review.avatar",
                    },
                    "children": Array [],
                    "code": Object {
                      "actions": Object {},
                    },
                    "properties": Object {},
                    "responsiveStyles": Object {
                      "large": Object {
                        "height": "30px",
                        "marginRight": "10px",
                        "width": "30px",
                      },
                    },
                    "tagName": "img",
                  },
                  Object {
                    "@type": "@builder.io/sdk:Element",
                    "actions": Object {},
                    "bindings": Object {},
                    "children": Array [
                      Object {
                        "@type": "@builder.io/sdk:Element",
                        "actions": Object {},
                        "bindings": Object {},
                        "children": Array [
                          Object {
                            "@type": "@builder.io/sdk:Element",
                            "bindings": Object {
                              "component.options.text": "review.author",
                            },
                            "component": Object {
                              "name": "Text",
                              "options": Object {},
                            },
                            "tagName": "span",
                          },
                        ],
                        "code": Object {
                          "actions": Object {},
                        },
                        "properties": Object {},
                        "tagName": "div",
                      },
                      Object {
                        "@type": "@builder.io/sdk:Element",
                        "actions": Object {},
                        "bindings": Object {},
                        "children": Array [
                          Object {
                            "@type": "@builder.io/sdk:Element",
                            "bindings": Object {
                              "component.options.text": "review.reviewMessage",
                            },
                            "component": Object {
                              "name": "Text",
                              "options": Object {},
                            },
                            "tagName": "span",
                          },
                        ],
                        "code": Object {
                          "actions": Object {},
                        },
                        "properties": Object {},
                        "tagName": "div",
                      },
                    ],
                    "code": Object {
                      "actions": Object {},
                    },
                    "properties": Object {},
                    "tagName": "div",
                  },
                ],
                "code": Object {
                  "actions": Object {},
                },
                "properties": Object {},
                "responsiveStyles": Object {
                  "large": Object {
                    "background": "white",
                    "borderRadius": "5px",
                    "boxShadow": "0 2px 5px rgba(0, 0, 0, 0.1)",
                    "display": "flex",
                    "margin": "10px",
                    "padding": "10px",
                  },
                },
                "tagName": "div",
              },
            ],
            "component": Object {
              "name": "Fragment",
            },
            "repeat": Object {
              "collection": "state.reviews",
              "itemName": "review",
            },
          },
        ],
        "code": Object {
          "actions": Object {},
        },
        "properties": Object {},
        "tagName": "div",
      },
    ],
    "jsCode": "var props = state;

Object.assign(state, { reviews: [], showReviewPrompt: false });

fetch(
  \`https://stamped.io/api/widget/reviews?storeUrl=builder-io.myshopify.com&apiKey=\${
    props.apiKey || \\"pubkey-8bbDq7W6w4sB3OWeM1HUy2s47702hM\\"
  }&productId=\${props.productId || \\"2410511106127\\"}\`
)
  .then((res) => res.json())
  .then((data) => {
    state.reviews = data.data;
  });
",
    "tsCode": "var props = state;

useState({ reviews: [], showReviewPrompt: false });

onMount(() => {
  fetch(
    \`https://stamped.io/api/widget/reviews?storeUrl=builder-io.myshopify.com&apiKey=\${
      props.apiKey || \\"pubkey-8bbDq7W6w4sB3OWeM1HUy2s47702hM\\"
    }&productId=\${props.productId || \\"2410511106127\\"}\`
  )
    .then((res) => res.json())
    .then((data) => {
      state.reviews = data.data;
    });
});
",
  },
}
`;

exports[`Builder Stamped 2`] = `
"import { Show, For } from \\"@jsx-lite/core\\";

export default function MyComponent(props) {
  onMount(() => {
    useState({
      reviews: [],
      showReviewPrompt: false,
    });
    onMount(() => {
      fetch(
        \`https://stamped.io/api/widget/reviews?storeUrl=builder-io.myshopify.com&apiKey=\${
          props.apiKey || \\"pubkey-8bbDq7W6w4sB3OWeM1HUy2s47702hM\\"
        }&productId=\${props.productId || \\"2410511106127\\"}\`
      )
        .then((res) => res.json())
        .then((data) => {
          state.reviews = data.data;
        });
    });
  });

  return (
    <div>
      <button onClick={(event) => (state.showReviewPrompt = true)}>
        <div>Write a review</div>
      </button>
      <Show when={state.showReviewPrompt}>
        <input placeholder=\\"Email\\" />
        <input
          placeholder=\\"Title\\"
          css={{
            display: \\"block\\",
          }}
        />
        <textarea
          placeholder=\\"How was your experience?\\"
          css={{
            display: \\"block\\",
          }}
        ></textarea>
        <button
          onClick={(event) => (state.showReviewPrompt = false)}
          css={{
            display: \\"block\\",
          }}
        >
          <div>Submit</div>
        </button>
      </Show>
      <For each={state.reviews}>
        {(review, index) => (
          <div
            css={{
              margin: \\"10px\\",
              padding: \\"10px\\",
              background: \\"white\\",
              display: \\"flex\\",
              borderRadius: \\"5px\\",
              boxShadow: \\"0 2px 5px rgba(0, 0, 0, 0.1)\\",
            }}
          >
            <img
              src={review.avatar}
              css={{
                height: \\"30px\\",
                width: \\"30px\\",
                marginRight: \\"10px\\",
              }}
            />
            <div>
              <div>
                <div>{review.author}</div>
              </div>
              <div>
                <div>{review.reviewMessage}</div>
              </div>
            </div>
          </div>
        )}
      </For>
    </div>
  );
}
"
`;
